File: C++/NumberOfZigzagArrays_I.cpp
#include<vector>
using namespace std;

class NumberOfZigzagArrays_I {
public:
    static constexpr int MOD = 1000000007;

    int zigZagArrays(int n, int l, int r) {
        int m = r - l + 1;

        vector<long long> up(m, 1), down(m, 1);
        vector<long long> pre(m + 1), suf(m + 1);
        vector<long long> newUp(m), newDown(m);

        for (int len = 2; len <= n; len++) {
            pre[0] = 0;
            for (int i = 0; i < m; i++) {
                pre[i + 1] = (pre[i] + down[i]) % MOD;
            }

            suf[m] = 0;
            for (int i = m - 1; i >= 0; i--) {
                suf[i] = (suf[i + 1] + up[i]) % MOD;
            }

            for (int i = 0; i < m; i++) {
                newUp[i] = pre[i];
                newDown[i] = suf[i + 1];
            }

            up.swap(newUp);
            down.swap(newDown);
        }

        long long ans = 0;
        for (int i = 0; i < m; i++) {
            ans = (ans + up[i] + down[i]) % MOD;
        }

        return (int)ans;
    }
};

File: Java/NumberOfZigzagArrays_I.java
import java.util.Arrays;

public class NumberOfZigzagArrays_I {
    int mod = 1_000_000_007;
    public int zigZagArrays(int n, int l, int r) {
        int m = r - l + 1;
        long[] up = new long[m];
        long[] down = new long[m];
        Arrays.fill(up, 1);
        Arrays.fill(down, 1);

        for (int i = 2; i <= n; i++) {
            long[] preDown = new long[m + 1];
            long[] suffUp = new long[m + 1];
            for (int j = 0; j < m; j++) {
                preDown[j + 1] = (preDown[j] + down[j]) % mod;
            }
            for (int j = m - 1; j >= 0; j--) {
                suffUp[j] = (suffUp[j + 1] + up[j]) % mod;
            }
            long[] newUp = new long[m];
            long[] newDown = new long[m];
            for (int j = 0; j < m; j++) {
                newUp[j] = preDown[j];
                newDown[j] = suffUp[j + 1];
            }

            up = newUp;
            down = newDown;
        }

        long ans = 0;
        for (int j = 0; j < m; j++) {
            ans = (ans + up[j] + down[j]) % mod;
        }

        return (int) ans;
    }

    public static void main(String[] args) { }
}